src/derivatives.jl

```diff
@@ -1,15 +1,53 @@
 function derivative(A, t, order = 1)
-    ((t < A.t[1] || t > A.t[end]) && !A.extrapolate) && throw(ExtrapolationError())
-    iguess = A.iguesser
+    (order ∉ (1, 2)) && throw(DerivativeNotFoundError())
+    if t < first(A.t)
+        _extrapolate_derivative_down(A, t, order)
+    elseif t > last(A.t)
+        _extrapolate_derivative_up(A, t, order)
+    else
+        iguess = A.iguesser
+        (order == 1) ? _derivative(A, t, iguess) :
+        ForwardDiff.derivative(t -> begin
+                _derivative(A, t, iguess)
+            end, t)
+    end
+end
 
-    return if order == 1
-        _derivative(A, t, iguess)
-    elseif order == 2
+function _extrapolate_derivative_down(A, t, order)
+    (; extrapolation_left) = A
+    typed_zero = zero(first(A.u) / one(A.t[1]))
+    if extrapolation_left == ExtrapolationType.none
```

Comment:
Slightly knit picky, but I think checking this one line above would prevent a few wasteful computation of `typed_zero`